File: python/pawn.py
import re

SPACE = " "
MAKE_EP_FROM = {"P": "5", "p": "4"}
MAKE_CAP_FROM = {"P": {"3": "2", "4":"3", "5": "4", "6": "5", "7": "6"},
                 "p": {"6": "7", "5": "6", "4": "5", "3": "4", "2": "3"}}
MAKE_PROM_FROM = {"P": "7", "p": "2"}
MAKE_REG_FROM = {"P": {"3": ["2"], "4": ["3", "2"],"5": ["4"],"6": ["5"],"7": ["6"]},
                 "p": {"6": ["7"], "5": ["6", "7"],"4": ["5"],"3": ["4"],"2": ["3"]}}

def make_pawn_move(move, board_view, piece_view):
    if is_ep(move, board_view):
        return make_ep(move, board_view, piece_view)
    elif is_capture(move):
        return make_capture(move, board_view, piece_view)
    elif is_promotion(move):
        return make_promotion(move, board_view, piece_view)
    elif is_regular_move(move):
        return make_regular_move(move, board_view, piece_view)

def is_ep(move, board_view):
    return re.fullmatch("[Pp][a-h]x[a-h][2-7]", move) is not None and board_view[move[-2:]] == SPACE

def is_capture(move):
    return re.fullmatch("[Pp][a-h]x[a-h][2-7]", move) is not None

def is_promotion(move):
    return re.fullmatch("[Pp][a-h](x[a-h])?[18][RNBQrnbq]", move) is not None

def is_regular_move(move):
    return re.fullmatch("[Pp][a-h][2-7]", move) is not None

def make_ep(move, board_view, piece_view):
    pawn, to_sq, from_file = move[0], move[-2:], move[1]
    from_sq = from_file + MAKE_EP_FROM[pawn]
    if pawn == "P":
        cap_at_sq = to_sq[0] + str(int(to_sq[1]) - 1)
    else:
        cap_at_sq = to_sq[0] + str(int(to_sq[1]) + 1)
    
    board_view[from_sq] = SPACE
    board_view[to_sq] = pawn
    board_view[cap_at_sq] = SPACE
    piece_view[pawn].append(to_sq)
    piece_view[pawn].remove(from_sq)
    piece_view[pawn.swapcase()].remove(cap_at_sq)

    return board_view, piece_view

def make_capture(move, board_view, piece_view):
    pawn, to_sq, from_file = move[0], move[-2:], move[1]
    from_sq = from_file + MAKE_CAP_FROM[pawn][to_sq[1]]
    captured_piece = board_view[to_sq]

    board_view[from_sq] = SPACE
    board_view[to_sq] = pawn
    piece_view[pawn].append(to_sq)
    piece_view[pawn].remove(from_sq)
    piece_view[captured_piece].remove(to_sq)

    return board_view, piece_view

def make_promotion(move, board_view, piece_view):
    pawn, to_sq, promoted_to = move[0], move[-3:-1], move[-1]
    if "x" in move:
        from_file = move[1]
        captured_piece = board_view[to_sq]
    else:
        from_file = to_sq[0]
        captured_piece = SPACE
    from_sq = from_file + MAKE_PROM_FROM[pawn]

    board_view[to_sq] = promoted_to
    board_view[from_sq] = SPACE
    piece_view[pawn].remove(from_sq)
    piece_view[promoted_to].append(to_sq)
    if captured_piece != SPACE:
        piece_view[captured_piece].remove(to_sq)

    return board_view, piece_view

def make_regular_move(move, board_view, piece_view):
    pawn, to_sq, from_file = move[0], move[-2:], move[-2]
    if len(MAKE_REG_FROM[pawn][to_sq[1]]) == 1:
        from_rank = MAKE_REG_FROM[pawn][to_sq[1]][0]
    else:
        for rank in MAKE_REG_FROM[pawn][to_sq[1]]:
            if board_view[from_file + rank] == pawn:
                from_rank = rank
                break
    from_sq = from_file + from_rank

    board_view[from_sq] = SPACE
    board_view[to_sq] = pawn
    piece_view[pawn].append(to_sq)
    piece_view[pawn].remove(from_sq)

    return board_view, piece_view

File: kotlin/src/Pawn.kt
val makeEpFrom = hashMapOf<String, String>("P" to "5", "p" to "4")

val makeCapFrom = hashMapOf("P" to hashMapOf<String, String>("3" to "2", "4" to "3", "5" to "4", "6" to "5", "7" to "6"),
    "p" to hashMapOf<String, String>("6" to "7", "5" to "6", "4" to "5", "3" to "4", "2" to "3"))

val makePromFrom = hashMapOf<String, String>("P" to "7", "p" to "2")

val makeRegFrom = hashMapOf("P" to hashMapOf<String, List<String>>("3" to listOf("2"), "4" to listOf("3", "2"),
        "5" to listOf("4"), "6" to listOf("5"), "7" to listOf("6")),
    "p" to hashMapOf("6" to listOf("7"), "5" to listOf("6", "7"), "4" to listOf("5"), "3" to listOf("4"),
            "2" to listOf("3")))

fun isEp(move: String, boardView: HashMap<String, String>): Boolean {
    return Regex("[Pp][a-h]x[a-h][2-7]").matchEntire(move) != null && boardView[move.slice(move.length-2 until move.length)] == SPACE
}

fun isCapture(move: String): Boolean {
    return Regex("[Pp][a-h]x[a-h][2-7]").matchEntire(move) != null
}

fun isPromotion(move: String): Boolean {
    return Regex("[Pp][a-h](x[a-h])?[18][RNBQrnbq]").matchEntire(move) != null
}

fun isRegularMove(move: String): Boolean {
    return Regex("[Pp][a-h][2-7]").matchEntire(move) != null
}

fun makeEp(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board{
    val pawn: String = move[0].toString()
    val toSq: String = move.slice(move.length-2 until move.length)
    val fromFile: String = move[1].toString()
    val fromSq: String = fromFile + makeEpFrom[pawn]
    val capAtSq: String
    if (pawn == "P") {
        capAtSq = toSq[0].toString() + (toSq[1].toString().toInt() - 1).toString()
    }
    else {
        capAtSq = toSq[0].toString() + (toSq[1].toString().toInt() + 1).toString()
    }

    boardView[fromSq] = SPACE
    boardView[toSq] = pawn
    boardView[capAtSq] = SPACE
    pieceView[pawn]!!.add(toSq)
    pieceView[pawn]!!.remove(fromSq)
    when(pawn) {
        "P" -> pieceView["p"]!!.remove(capAtSq)
        else -> pieceView["P"]!!.remove(capAtSq)
    }

    return Board(boardView, pieceView)
}

fun makeCapture(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    val pawn: String = move[0].toString()
    val toSq: String = move.slice(move.length - 2 until move.length)
    val fromFile: String = move[1].toString()
    val fromSq = fromFile + makeCapFrom[pawn]!![toSq[1].toString()]
    val capturedPiece = boardView[toSq]
    boardView[fromSq] = SPACE
    boardView[toSq] = pawn
    pieceView[pawn]!!.add(toSq)
    pieceView[pawn]!!.remove(fromSq)
    pieceView[capturedPiece]!!.remove(toSq)

    return Board(boardView, pieceView)
}

fun makePromotion(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    val pawn: String = move[0].toString()
    val toSq: String = move.slice(move.length - 3 until move.length - 1)
    val promotedTo: String = move[move.length - 1].toString()
    val fromFile: String
    val capturedPiece: String
    if (move.contains("x")) {
        fromFile = move[1].toString()
        capturedPiece = boardView[toSq].toString()
    } else {
        fromFile = toSq[0].toString()
        capturedPiece = SPACE
    }
    val fromSq = fromFile + makePromFrom[pawn]

    boardView[toSq] = promotedTo
    boardView[fromSq] = SPACE
    pieceView[pawn]!!.remove(fromSq)
    pieceView[promotedTo]!!.add(toSq)
    if (capturedPiece != SPACE) {
        pieceView[capturedPiece]!!.remove(toSq)
    }
    return Board(boardView, pieceView)
}

fun makeRegularMove(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    val pawn: String = move[0].toString()
    val toSq: String = move.slice(move.length - 2 until move.length)
    val fromFile: String = move[move.length - 2].toString()
    var fromRank: String = SPACE
    if ((makeRegFrom[pawn]!![toSq[1].toString()])!!.size == 1) {
        fromRank = makeRegFrom[pawn]!![toSq[1].toString()]!![0]
    } else {
        for (rank in makeRegFrom[pawn]!![toSq[1].toString()]!!) {
            if (boardView[fromFile + rank] == pawn) {
                fromRank = rank
                break
            }
        }
    }
    val fromSq = fromFile + fromRank

    boardView[fromSq] = SPACE
    boardView[toSq] = pawn
    pieceView[pawn]!!.add(toSq)
    pieceView[pawn]!!.remove(fromSq)

    return Board(boardView, pieceView)
}

fun makePawnMove(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    when {
        isEp(move, boardView) -> {
            return makeEp(move, boardView, pieceView)
        }
        isCapture(move) -> {
            return makeCapture(move, boardView, pieceView)
        }
        isPromotion(move) -> {
            return makePromotion(move, boardView, pieceView)
        }
        isRegularMove(move) -> {
            return makeRegularMove(move, boardView, pieceView)
        }
    }
    return Board(boardView, pieceView)
}